### PR TITLE
Generating expr

### LINKED ISSUE #154: Arbitrary instance for HashMap is unfortunate
`arbitrary` for `HashMap k v` (in `tests/Strictness.hs`) is based on `fromList`. This means that only `HashMap` shapes that can be produced by `fromList` will be represented in the inputs to test functions.

### PR DESCRIPTION
**Pull request by phadej**
This is try to fix #154 

We could rewrite all `[Key]` or `[(Key, Int)]` based properties into `Expr Key ()` or `Expr Key Int` based ones, respectively. I think including `union` is worth it, as it is has non-trivial implementation.

First commit is from #156.

### CODE REVIEW

File: tests/Utils.hs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  deriving Show
+
+instance (Arbitrary k, Arbitrary v) => Arbitrary (Expr k v) where
+    arbitrary = sized arb
+      where
+        arb n | n <= 0 = oneof leafs
+        arb n = oneof $ leafs ++
+            [ liftA3 ExprInsert arbitrary arbitrary (arb (n - 1))
+            , liftA2 ExprDelete arbitrary (arb (n - 1))
+            , liftA2 ExprUnion (arb (n `div` 2)) (arb (n - n `div` 2))
```
Comment (treeowl) [inline tests/Utils.hs]: Why `div` and not `quot`?

Comment (phadej) [inline tests/Utils.hs]:
```
Prelude Test.QuickCheck> 3 `div` 2 + 3 `quot` 2
2
```

**Comment by phadej** [comment]
> I'm not quite convinced this fully solves the problem. We want to guarantee (as best we can) that every valid HashMap tree is available to the Arbitrary instance. Do we have good reasons to believe that all such trees can be generated by this language?

I don't know, I looked at non-trivial `-> HashMap k v` functions in ` Base.hs` module.

> An alternative, more complicated, approach (but quite possibly the right one) would be to use an expression language that includes all primitive operations, interpreting it using HashMap or HashSet and also (say) Data.Map or Data.Set.

AFAICS the `insert`, `delete` and `union` are a good start.

> One of the key ideas of the containers test suite is that all the invariants for each data structure are expressed in a single valid function. In this context, that would allow your Expr interpretation to find the smallest invalid subexpression, rather than waiting for that invariant failure to show up as a hard-to-debug semantic error.

That's a good idea, but it can be added independently. I'm not 100% sure what are the invariants.

**Comment by treeowl** [comment]
Can you rebase this, please?

**Comment by phadej** [comment]
rebased

**Comment by phadej** [comment]
ping @treeowl, is there something I can do to get this off my "open PR" -list? :)

**Comment by treeowl** [comment]
Uh, I seem to have forgotten about this. I'll try to get to it in the next couple days.

**Comment by phadej** [comment]
This seems to be still forgotten...

**Comment by treeowl** [comment]
I think we've removed the need for it by removing the representation
redundancy. See the changes in Eq implementation.

**Pull request closed without merging**